### PR TITLE
Make figures reproducible using Docker

### PR DESCRIPTION
**Pull request by fabianegli**
## This PR adds

- an R script to install the dependencies to reproduce the graphics in figure 2e
- a `Dockerfile` to create a Docker image which has all the dependencies installed
- an additional section in the README.md file which explains how to recreate the graphics using this Docker image
- slightly adapts the R script `make_figure2e.R` for non-interactive reproduction of the graphics

## This PR does not add

- documentation of the installed version of packages
- any link to a ready built Docker image

## Thoughts

- I could upload an image to my individual dockerhub account, but am not sure this is desirable?
- The code for Figure 3 does not yet live in this repo, but a fork. Thus, this PR does not take figure 3 into account.

I hope this is of any use to you.

NB: This is close to the first working version of a cross-platform reproduction method for this code I came up with and has no aspirations to be minimal in terms of installed R packages or the size of the resulting Docker image.

## UPDATE

This PR absolutely adds a link to a ready built Docker image to the README.md

### CODE REVIEW

**Comment by cvanderaa** [comment]
Hello @fabianegli ! 

Thanks a lot for this great PR and taking the time of setting up a Docker file. I'll merge your PR in this fork (and then forward it to the main repo), but first I would like to discuss your thoughts:

> I could upload an image to my individual dockerhub account, but am not sure this is desirable?

- I think it is desirable to have a perennial solution for getting the Docker image. If you could share it from your DockerHub account, that would be great. Another solution I see, is to use a specific tag when calling the Bioconductor docker image, so using 

```
FROM bioconductor/bioconductor_docker:RELEASE_3_16
```
- Also, can I suggest to rename your image `scp-proteomics-docker-image`? In SCP, the "P" already stands for proteomics. I prefer a name that is closer to the name of the GitHub repo, like `scp-recommendations-docker-image`.

>The code for Figure 3 does not yet live in this repo, but a fork. Thus, this PR does not take figure 3 into account.

I'm not sure to which figure 3 you are referring to. In the preprint and the Nature methods paper, figure 3 is an illustration that was manually generated using a vector graphics editor.

**Comment by fabianegli** [comment]
Thank you for embracing the PR and your feedback.

I agree that it is desirable to have a docker image, I was just not sure if you'd appreciate it being offered via my dockerhub account. But if that is fine with you, I will gladly set this up.

I am more than happy to rename the image with a more specific and descriptive name. I added the "-docker-image" part so that people unfamiliar with docker have a better idea of what the command does, but this can arguably be removed, too. Another option could be to add the year, to somewhat mimic a citation. e.g. "scp-recommendations-2022" or even ""scp-recommendations-gatto-2022". What do you think?

Next steps:

- [x] decide on the image name
- [x] create/upload the image
- [x] document the image availability and workflow using it in the README.md

About figure 3 I think I was a bit confused because I saw the commit message ["Added source file for Figure3"](https://github.com/SlavovLab/SCP_recommendations/commit/c4846cb1625dd3f93d2f9f32325adf4350b48dca) and assumed source was referring to R code, but it is a SVG. Just saw it in the Network and didn't look at it in detail, so that's on me 🙃

**Comment by fabianegli** [comment]
Small sidenote on the version documentation: versioning the "FROM" image is a good step towards it, but I think I remember the maintainers sometimes backporting *fixes* to versioned containers. See https://rocker-project.org/use/reproducibility.html for reference. And this would still not be sufficient because the packages installed with the installation script also need to be versioned. Once the image is available, the installed versions can be found by inspecting/running the image, but it would still be nice to document a snapshot in a more easily accessible format like writning the output of `sessionInfo()` into a file, but there might be better options, too. Like `renv`, but I am not versed enough to know if that would be warranted in this case.

**Comment by fabianegli** [comment]
There is one more question: The graphics produced by the container are actually not exactly the same in that it makes bigger dots than what is seen in the paper. I wonder if this should be somehow addressed?

**Comment by cvanderaa** [comment]
Sorry for the delay... 

- For the naming, I personally like `scp-recommendations-2022-docker`
- About figure 3: it's my bad, I shouldn't have polluted the git history with the SVG file.
- Regarding backporting fixes: I agree with you that even starting from a versioned images does not provide a 100% guarantee for reproducibility. I think Bioconductor also occasionally  performs fixes on the latest stable image. I already tried to play with `renv` but it does not solve the Bioconductor problem, because once a package is fixed, its version is incremented and the previous version is no longer available for `renv` to download... Bottom line, the best is that you provide open access to the Docker image you created; this will provide 100% guarantee of reproducibility of the current code. Providing a reproducible and documented `DockerFile` will facilitate generating future images and we should release them as a new tag (or new DockerHub repo) upon improvement of the code in this GitHub repo.
- Regarding the dot size: my answer would depend on the difference of size. Are the dots so big that we lose information? If so, then I suggest we adress this as a separate issue and I can have a look once you share the Docker image. If not, then I would say it is overkill.

**Comment by fabianegli** [comment]
Thank you :-)

- For the name: I'll go with `scp-recommendations-2022` because the `docker` seems redundant.
- No worries about fig 3. It's no big deal.
- I uploaded an image and tested it. The PR has been updated accordingly.

I'll attach a screenshot of the two versions of 2e_A (left: original; right: reproduced):

Screenshot 2023-03-21 at 15 20 53

NB: this PR has some level of inception... the comand in the readme won't work out of the box as long as the PR isn't merged yet and requires the replacement of the `SlavovLab` GitHub user part of the URL to mine (fabianegli) in the command to clone the repo.

**Comment by fabianegli** [comment]
Just in case anyone is looking for the command used to create the image:

```
docker build --output=type=registry,push=true --tag fabianegli/scp-recommendations-2022:1.0 .
```

Run from within this repo at https://github.com/cvanderaa/SCP_recommendations/pull/1/commits/61f616f2c4f976f19869dfcdb6da9bad0d2408ea approximately 20 minutes before the upload of the images.

**Comment by cvanderaa** [comment]
Thanks a lot for uploading the image and adapting the README! After a quick test, it seems that the difference in point size is linked to different default settings of `save_image` (and `ggsave`) probably linked to the underlying graphics library. I could get something similar to the original figure when using: 

```r
save_image(ply, plyFile, width = 1000, height = 1000)
```

Feel free to adapt it, in my opinion, it's a detail. 

Many thanks for your efforts, let me know when you're ready for me to merge the PR.

File: code/make_figure2e.R [hunk 1]
```diff
@@ -135,9 +136,9 @@ ply <- plot_ly(x = x, y = y, z = z,
                                       up = list(x = 0,
                                                 y = 0,
                                                 z = 1))))
-ply
 plyFile <- "figs/figure2e_A.png"
-save_image(ply, normalizePath(plyFile))
+reticulate::py_run_string("import sys")
+save_image(ply, plyFile)
```
Comment (fabianegli) [inline code/make_figure2e.R:141]:
```suggestion
save_image(ply, plyFile, width = 1000, height = 1000)
```

**Comment by fabianegli** [comment]
With pleasure!

Thank you for experimenting with the plot. I don't know about the reason, it might very well be some changed defaults. I'd prefer to keep the figure size constant and instead change the dot size to "5/3" which makes the plot almost a pixel perfect match (and also keeps the three in there).

I'd be happy to have it merged the way it is now 😄

**Comment by cvanderaa** [comment]
Again thanks a lot for the PR :pray:  I'll forward it to the main repository.

**Pull request merged**